Extend Nanoevents vector and add tests

jrueb opened the pull request:
Closes #445

Besides adding the features mentioned in #445, this also adds a `negative` mixin, allowing thngs like `b = -a`. Also added `PolarTwoVector.multiply` and `SphericalThreeVector.multiply`.
Additionally I fixed the `PtEtaPhiMLorentzVector.multiply` and `PtEtaPhiELorentzVector.multiply` functions, which gave correct results only for non-negative factors.
This also adds a number of unit tests for the vectors.
Finally I used @nsmith- suggestion for naming `pvec`, but this can still easily be adjusted.

jrueb committed:
Extend Nanoevents vector and add tests

Closes #445

nsmith- committed:
Merge branch 'master' into extendvector

jrueb committed:
Fix typo in documentation

nsmith- committed:
Merge branch 'master' into extendvector

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -136,13 +152,30 @@ def multiply(self, other):
         )
 
+    @awkward.mixin_class_method(numpy.divide, {numbers.Number})
+    def divide(self, other):
```
Could this be implemented by calling multiply on the reciprocal?

jrueb replied:
Would this not be a performance loss? `other` can be an array of numbers, thus it would first allocate an array for the reciprocal and then another one for the multiplication. And I could imagine reciprocal & multiplication also takes more cycles than just a division.

nsmith- replied:
In general multiplication is much cheaper than division. So by computing the reciprocal and re-using it for each cartesian component multiplication should actually increase performance. The extra memory cost can be avoided by overwriting, e.g. `other = 1 / other`

nsmith- replied:
I'm not sure how reference lifecycle behaves in a call stack, but python may do "the right thing" if you just implement this as `self.multiply(1 / other)`

jrueb replied:
The only difference between any of the `divide` methods would then by their documentation string. As everything inherits from `TwoVector` it would be enough to implement it only there. Do you want to keep the individual implementations?

nsmith- replied:
Sure, with a generic enough docstring just the TwoVector impl. is fine

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -189,6 +222,42 @@ def r2(self):
+        )
+
+    @awkward.mixin_class_method(numpy.divide, {numbers.Number})
+    def divide(self, other):
```
Same question about multiply with reciprocal

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -271,6 +356,34 @@ def multiply(self, other):
         )
 
+    @awkward.mixin_class_method(numpy.divide, {numbers.Number})
+    def divide(self, other):
```
Reciprocal

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -322,6 +435,48 @@ def p2(self):
+        )
+
+    @awkward.mixin_class_method(numpy.divide, {numbers.Number})
+    def divide(self, other):
```
Reciprocal?

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -406,6 +574,19 @@ def multiply(self, other):
         )
 
+    @awkward.mixin_class_method(numpy.divide, {numbers.Number})
+    def divide(self, other):
```
Reciprocal?

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -417,6 +598,75 @@ def delta_r(self, other):
+        )
+
+    @property
+    def boostvec(self):
```
If `t <= rho` (or equivalently `mass <= 0`) do we want to make the output NaN since there is no physical way to get to that reference frame?

jrueb replied:
I think this would make sense, but would you then also return NaN for zero vectors (there we have `0 == t == rho`)?

nsmith- replied:
As `t` approaches `rho`, `boostvec` approaches the unit vector (i.e. speed of light), so I think it would be consistent if we returned `self.pvec * numpy.minimum(1 / self.rho, 1 / self.t)`.

nsmith- replied:
I totally ignored your zero vector comment! Let me try again: `self.pvec * awkward.where(self.rho == 0, 0, numpy.minimum(1 / self.rho, 1 / self.t))`. I would make `self.rho` a temporary though as it may be expensive in some cases and is used twice.

lgray replied:
Well - unphysical lorentz transformations still belong to SO(1,3), instead of SO+(1,3). So it's still a lorentz xform but it cannot be described by a boost and a rotation. 

It may be better, and more correct, to simply warn that a boost was calculated outside of the restricted lorentz group. Then also to protect against zeros in some reasonable way.

This may also suggest returning the the boost tensor rather than just the vec... Or extending what we do to include that notion. It then becomes much easier to impose notions of physicality since it's pretty easy to check the lorentz xform tensor for what type it is.

jrueb replied:
I made it depend on `|t| <= rho`. It not returning a normal boost vector even if `|t| >> rho` sounds like a possible pitfall to me and one can easily get into this situation by `(-lorentzvec).boostvec`, which works fine in uproot_methods and ROOT. 
If the user tries to boost with a vector coming from `|t| <= rho` there will be a warning in the `boost` function.

nsmith- replied:
Not to wax too long on this but based on https://en.wikipedia.org/wiki/Lorentz_group#Connected_components I think actually improper transforms are outside SO(1, 3). But also |boostvec| > 1 would give complex-valued output in the current `boost` implementation so I'm not sure it makes any sense even as a parity-reversing transform.

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -584,14 +834,43 @@ def mass2(self):
-                "eta": self.eta,
-                "phi": self.phi,
-                "mass": self.mass * other,
+                "pt": self.pt * abs(other),
```
Making `abs(other)` a temporary would be a slight performance boost

nsmith- reviewed (changes requested):
Thanks a lot for this! Also for the needed tests. Perhaps @henryiii might want to fold them into the scikit-hep/vector unit tests when the time comes.

nsmith- commented on coffea/nanoevents/methods/vector.py:
```diff
@@ -417,6 +598,75 @@ def delta_r(self, other):
+        a vector into its own rest frame, one needs to use the negative of its `boostvec`
+        """
+        b2 = other.rho2
+        gamma = (1 - b2) ** (-0.5)
```
We should guard somehow against the case `b2 == 1` as well as `b2 == 0`.

jrueb replied:
Right now for `b2==1` there will be a warning and `inf` as vector components. Is there anything better result in this situation?

nsmith- replied:
That seems reasonable, I wasn't sure what the behavior was currently.

jrueb committed:
Nanoevents vector changes as discussed in the PR

Single implementation for division
Handle t < rho in boostvec
Use more temporary variables

nsmith- reviewed (approved):
Looks good
